### PR TITLE
Add/websocket

### PR DESCRIPTION
**Pull request by AHReccese**
#### Reference Issues/PRs
#41 
In this PR, the ability to use websocket as the communication medium for both client and server side is implemented. but as you know, websocket brings asynchronous context to the board and we do have the ability to let user define for example 3 ml models and push the fit function to the server and wait till get them done. but we need to design a proper mechanism for this purpose. right now it the underlying medium is using websocket and the point is for example that there is one connection instead of connection creation per request to server (like REST protocol).

### CODE REVIEW

**Comment by AHReccese** [comment]
All checks are passed, but for weird reason, the windows python 3.11 fails, as you can see in the attached picture, all tests are passed but the process gets killed, I will re-run this failed job in another time.

image

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,9 +6,19 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `validate_websocket_url` in `streaming.util.py`
+- `ML Streaming` WebSocket testcases
+- `ClientCommunicator` Enum in `streaming.communicator.py`
+- `ServerCommunicator` Enum in `streaming.communicator.py`
+- `WebSocketClientCommunicator` class in `streaming.communicator.py`
+- `WebSocketServerCommunicator` class in `streaming.communicator.py`
 - PyMilo exception types added in `pymilo/exceptions/__init__.py`
 - PyMilo exception types added in `pymilo/__init__.py`
 ### Changed
+- `PyMiloClient` updated to supprot protocol selection
```
Comment (sepandhaghighi) [inline CHANGELOG.md:19]: `client_communicator` parameter added to `PyMiloClient` class

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,9 +6,19 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `PyMiloServer` updated to support protocol selection
```
Comment (sepandhaghighi) [inline CHANGELOG.md:20]: `client_communicator` parameter added to `PyMiloServer` class

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]
Please use lowercase letters for the starting of docstring `:param` and `:type` sections. 

For example:

```
Handle WebSocket disconnection.
:param websocket: the WebSocket connection to remove.
:type websocket: WebSocket
```

File: CHANGELOG.md (hunk 2 above)
Comment (AHReccese) [reply CHANGELOG.md:20]: *server_communicator :))

File: pymilo/streaming/communicator.py [hunk 3]
```diff
@@ -188,3 +196,290 @@ def parse(self, body):
...
+                self._ps._encryptor.decrypt(message)
+            )
+        )
+
+    def run(self):
+        """Run the internal FastAPI server."""
+        uvicorn.run(self.app, host=self.host, port=self.port)
+
+
+class ClientCommunicator(Enum):
```
Comment (sepandhaghighi) [inline pymilo/streaming/communicator.py:474]: Here, you are overwriting the `interfaces.ClientCommunicator` class. Please choose another name for this enum.

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

File: pymilo/streaming/communicator.py (hunk 3 above)
Comment (AHReccese) [reply pymilo/streaming/communicator.py]: @sepandhaghighi 
Great point, thank you
I renamed it to `ClientCommunicationProtocol`

**Comment by AHReccese** [comment]
I will update `README.md` in a separate PR.
@sepandhaghighi

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,9 +6,19 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 ### Added
+- `validate_http_url` in `streaming.util.py`
```
Comment (sepandhaghighi) [inline CHANGELOG.md:9]: `validate_http_url` function in `streaming.util.py`

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,9 +6,19 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `validate_websocket_url` in `streaming.util.py`
```
Comment (sepandhaghighi) [inline CHANGELOG.md:10]: `validate_websocket_url` function in `streaming.util.py`

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,9 +6,19 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `ML Streaming` WebSocket testcases
+- `ClientCommunicationProtocol` Enum in `streaming.communicator.py`
```
Comment (sepandhaghighi) [inline CHANGELOG.md:12]: `CommunicationProtocol` Enum in `streaming.communicator.py`

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,9 +6,19 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `ServerCommunicationProtocol` Enum in `streaming.communicator.py`
```
Comment (sepandhaghighi) [inline CHANGELOG.md:13]: Remove this line

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,9 +6,18 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `validate_http_url` function in `streaming.util.py`
+- `validate_websocket_url` function in `streaming.util.py`
+- `ML Streaming` WebSocket testcases
+- `CommunicationProtocol` Enum in `streaming.communicator.py`
+- `WebSocketClientCommunicator` class in `streaming.communicator.py`
+- `WebSocketServerCommunicator` class in `streaming.communicator.py`
 - PyMilo exception types added in `pymilo/exceptions/__init__.py`
 - PyMilo exception types added in `pymilo/__init__.py`
 ### Changed
+- `client_communicator` parameter added to `PyMiloClient` class
```
Comment (sepandhaghighi) [inline CHANGELOG.md:18]: `communication_protocol` parameter added to `PyMiloClient` class

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,9 +6,18 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `server_communicator` parameter added to `PyMiloServer` class
```
Comment (sepandhaghighi) [inline CHANGELOG.md:19]: `communication_protocol` parameter added to `PyMiloServer` class

File: CHANGELOG.md [hunk 10]
```diff
@@ -6,9 +6,18 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- ML Streaming testcases updated to support protocol selection
```
Comment (sepandhaghighi) [inline CHANGELOG.md:20]:
```
`ML Streaming`
```

File: dev-requirements.txt [hunk 11]
```diff
@@ -5,6 +5,7 @@ uvicorn==0.32.0
 fastapi==0.115.2
 requests==2.32.3
 pydantic>=1.5.0
+websockets>=9.0
```
Comment (sepandhaghighi) [inline dev-requirements.txt:8]: `websockets==9.0`

`pydantic==1.5.0`

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

**Review by sepandhaghighi — APPROVED** [review]
@AHReccese 
Everything looks good to me for now 💯 
But remember that we need to manage the versions of `pydantic` and `scipy` in the future. If we cannot control their latest versions in the `dev-requirements.txt`, we should remove them from PyMilo requirement files.

**Pull request merged**